hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/signature/StringToSignProducer.java

```diff
@@ -249,6 +249,7 @@ private static String urlEncode(String str) {
     try {
       return S3Utils.urlEncode(str)
           .replaceAll("\\+", "%20")
+          .replaceAll("\\*", "%2A")
           .replaceAll("%7E", "~");
     } catch (UnsupportedEncodingException e) {
```

Comment:
Thanks @symious for catching is problem and submitting this fix.  As the uri is included in the string2sign and signature validation do we need to extend this encoding to support other special characters that need to be in ASCII hex or vice-versa?
For the URLEncoder class, all alphanumeric characters remain the same after encoding.  However the spaces are converted to '+' that we correct for in line 251.  All special characters ".", "-", "*", and  remain the"_"  same.  With all others converted to hex.  Are there any other cases of special characters that we need to convert to hex similar to the '*"?

Reply:
Also, is there an alternative encoder that removes the need for correcting a few characters after encoding?

Reply:
There are some SAFE Characters mentioned here: https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html.
Maybe we need also to check the availabilities for them?

Reply:
@symious thanks.  I checked with aws `botocore` for encoding the URI for the signature.  It looks like the `StringToSignProducer` encoder is equivalent now to aws `botocore`.  Which should be fine.  It is a python implementation that uses `urllib`.`quote`.
https://github.com/boto/botocore/blob/f2b0dbb800b8dc2a3541334d5ca1190faf900150/botocore/auth.py#L362

Within quote its safe characters are : Letters, digits, and the characters '_.-~'  as well as user added '/\~' .
https://docs.python.org/3/library/urllib.parse.html

If we are compatible with this then we should not have any unexpected problems.
With  the java class `java.net.URLEncoder,` the safe characters include in addition to the urllib.parse.quote safe characters  just "*".   Please check this as well.
https://docs.oracle.com/javase/1.5.0/docs/api/java/net/URLEncoder.html

Reply:
LGTM

Reply:
@neils-dev Thanks for the check. Agreed that we only need to add an aditional check for "*".

FYI: I also find the safe characters here based on your information. https://github.com/boto/botocore/blob/f2b0dbb800b8dc2a3541334d5ca1190faf900150/botocore/auth.py#L251